homeassistant/components/esphome/dashboard.py

```diff
@@ -4,69 +4,153 @@
 import asyncio
 from datetime import timedelta
 import logging
+from typing import Any
 
 import aiohttp
 from awesomeversion import AwesomeVersion
 from esphome_dashboard_api import ConfiguredDevice, ESPHomeDashboardAPI
 
 from homeassistant.config_entries import SOURCE_REAUTH, ConfigEntryState
 from homeassistant.const import EVENT_HOMEASSISTANT_STOP
-from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.core import CALLBACK_TYPE, Event, HomeAssistant, callback
 from homeassistant.helpers.aiohttp_client import async_get_clientsession
-from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
 from .const import DOMAIN
 
-KEY_DASHBOARD = "esphome_dashboard"
+_LOGGER = logging.getLogger(__name__)
 
 
-@callback
-def async_get_dashboard(hass: HomeAssistant) -> ESPHomeDashboard | None:
-    """Get an instance of the dashboard if set."""
-    return hass.data.get(KEY_DASHBOARD)
+KEY_DASHBOARD_MANAGER = "esphome_dashboard_manager"
 
+STORAGE_KEY = "esphome.dashboard"
+STORAGE_VERSION = 1
 
-async def async_set_dashboard_info(
-    hass: HomeAssistant, addon_slug: str, host: str, port: int
-) -> None:
-    """Set the dashboard info."""
-    url = f"http://{host}:{port}"
 
-    if cur_dashboard := async_get_dashboard(hass):
-        if cur_dashboard.addon_slug == addon_slug and cur_dashboard.url == url:
-            # Do nothing if we already have this data.
-            return
-        # Clear and make way for new dashboard
-        await cur_dashboard.async_shutdown()
-        del hass.data[KEY_DASHBOARD]
+async def async_setup(hass: HomeAssistant) -> None:
+    """Set up the ESPHome dashboard."""
+    # Try to restore the dashboard manager from storage
+    # to avoid reloading every ESPHome config entry after
+    # Home Assistant starts and the dashboard is discovered.
+    await async_get_or_create_dashboard_manager(hass)
+
 
-    dashboard = ESPHomeDashboard(hass, addon_slug, url, async_get_clientsession(hass))
-    try:
+@callback
+def async_get_dashboard_manager(hass: HomeAssistant) -> ESPHomeDashboardManager | None:
```

Comment:
This function is only used once inside the `get_dashboard` function, should it exist as a standalone function ?

Reply:
8197a2f3398292213ea1eb4d2e6f64edfe4c420c